### PR TITLE
Adds new non-parameterized activation functions, adds momentum and LR decay for SGD

### PR DESCRIPTION
**Pull request by syaffers**
Saw your post on reddit, so I thought I'd try my hands at implementing some of the optimization techniques for SGD and also some activation functions. I tried to keep the code as close to the original structure as I can (context variable names are different in the "signature" of the activation functions). Let me know if you need more of this, I'm keen to contribute.

### CODE REVIEW

**Comment by kaszperro** [comment]
Hi, thanks for this pull request, it looks decent. I changed the autograd structure a bit though. It's forward and backward methods aren't static any more. I decided to do it, because now it's possible to specify some unique properties for each autograd object, like in new Reshape() operation, where you can specify new shape in constructor. I added also method for saving informations for backprop, which should be used, because it allows to perform some checks in the future.
Please fix it, and I will merge your pull request :).

**Comment by syaffers** [comment]
Noted. Can I find the new updates to the autograd function in the development branch? I should have started there tbh, only realized that when I was making a pull request.

**Comment by kaszperro** [comment]
Master and development branches should be even now. I've realised that I've committed unfinished Conv2D layer and small bug in Reshape to the master branch.

**Comment by syaffers** [comment]
Updated the code, should be good now. I think there should be a way to unit test the activations rather than trying to see if it works by running training. Could be useful for future commits.

File: deepy/autograd/optimizers.py [hunk 1]
```diff
@@ -18,14 +18,35 @@ def step(self):
...
+
+            if variable not in self.state:
+                self.state[variable] = {}
+
+            state = self.state[variable]
+
+            if len(state) == 0:
+                self.initialize_state(state, variable)
+
+            state['v'] = self.mu * state['v'] - self.lr * variable.grad
```
Comment (kaszperro) [inline deepy/autograd/optimizers.py]:
```suggestion
            state['v'] = - self.mu * state['v'] - self.lr * variable.grad
```
Shouldn't be there minus sign?

Comment (kaszperro) [reply deepy/autograd/optimizers.py]: Silly me, yours code is of course correct.

**Comment by kaszperro** [comment]
Good job, keep it up :). 
I'll add unit tests, it'd simplify adding new futures. 
I think that now, main focus would be creating Conv2D and after that decent comments and documentation.

**Comment by syaffers** [comment]
Great! I'm thinking of adding the parameterized activations (LeakyReLU, SELU, etc.) next.
While it's still manageable, I thinking of commenting parts of the code. Are you thinking about using reStructured text or just plain docstrings?
Sorry for the late reply, was offline for a bit.

**Comment by kaszperro** [comment]
I think that https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html is good commenting pattern. From what I've seen, PyTorch also uses this type of comments.

**Pull request merged**